CHANGELOG_PENDING.md

```diff
@@ -44,6 +44,8 @@ Special thanks to external contributors on this release:
 
 ### IMPROVEMENTS
 
+- [internal/autofile] [#7333](https://github.com/tendermint/tendermint/pull/7333) autofile conservatively open its destination file and no longer opens a file every second. Only if the prior file doesn't exist does it create a fresh one. It however invokes .Sync every 1 second. This improves performance and reduces on file resources consumed.
```

Comment:
I'm a little worried about parts of this change. I really like the lazy creation of files rather than (potentially) creating empty files that are never (or rarely?) used. It might not help very much (the case where there isn't much load, seems like the primary case where we might be able skip file creation, which might make benchmarks faster but is unlikely to actually impact the performance of the system in any meaningful way.

I'm also a little bit worried about forcing automatic fsyncs more frequently, mostly because they can block unexpectedly under larger loads and can disrupt the otherwise well ordered function of the IO scheduler. Some times you have to force an fsync, and there are lots of good reasons for that, but I'm not sure this is one. The autofile tool is used by the WAL in the consensus reactor, and the wal implementation has it's own background process which periodically forces fsyncs (every 2 seconds, by default). In addition, the consensus machinery will trigger fsyncs periodically in response to behavior in the system. Adding more could collide or block these other calls, and that risk seems untested.

All of the concerns not withstanding, I'm a little worried about the assertions in this change log entry: I think we don't need to outline all of the behavioral changes here. "Changed behavior to allow lazy file creation with more regular fsync to rationalize resource creation." I'm a little hesitant to make assertions about performance improvement without more data.

Reply:
>I'm a little worried about parts of this change. I really like the lazy creation of files rather than (potentially) creating empty files that are never (or rarely?) used. It might not help very much (the case where there isn't much load, seems like the primary case where we might be able skip file creation, which might make benchmarks faster but is unlikely to actually impact the performance of the system in any meaningful way.

I don't understand what you mean but "unlikely to actually impact the performance of the system in any meaningful way", this change is the result of a clear exhibit of too many files being created every second and causing "too many files" open problems in tharsis/evmosd and other chains, and well instrumented in the Go standard library to show much less files created; essentially 1 file for as long as necessary instead of 86,400 files every 24 hours.

>I'm also a little bit worried about forcing automatic fsyncs more frequently, mostly because they can block unexpectedly under larger loads and can disrupt the otherwise well ordered function of the IO scheduler. 

It is no different from before, and in fact before we had .Sync() and .Close() every 1 second, now we have only .Sync() so this concern I don't think applies to this change.

>In addition, the consensus machinery will trigger fsyncs periodically in response to behavior in the system. Adding more could collide or block these other calls, and that risk seems untested.

The prior code did exactly .Sync() which is inside .Close() then a .Open afterwards, am not sure this concern stands.

>I'm a little worried about the assertions in this change log entry... I'm a little hesitant to make assertions about performance improvement without more data.

I showed the changes to file creation having instrumented the Go runtime, I kindly implore you to independently verify them yourself too.

Reply:
> I don't understand what you mean but "unlikely to actually impact the performance of the system in any meaningful way", this change is the result of a clear exhibit of too many files being created every second and causing "too many files" open problems in tharsis/evmosd and other chains, and well instrumented in the Go standard library to show much less files created; essentially 1 file for as long as necessary instead of 86,400 files every 24 hours.

The "too many open files" error happens when there are too many *concurrently* opened files, (e.g. files that we're opening and not closing,) and so I would believe that this change really wouldn't impact things assuming that we are actually closing files. 

I think I would feel more comfortable about this change, if we had more systemic benchmarks of:
- the impact of WAL file size on replay?
- the impact of mixing WAL events from different blocks?
- the impact of having a WAL implementation that isn't persisted to disk vs the current implementation vs this change. 

> I showed the changes to file creation having instrumented the Go runtime, I kindly implore you to independently verify them yourself too.

Sorry I missed this. Is there a place where I can isolate this change?